osquery/tables/system/darwin/keychain_acl.cpp

```diff
@@ -0,0 +1,334 @@
+                             keychain, CSSM_DL_DB_RECORD_ANY, NULL, &search););
+  if (os_status != 0) {
+    return Status(os_status,
+                  "Could not pull keychain items from the search API");
+  }
+  if (search == nullptr) {
+    return Status(1, "keychain search object was not populated properly");
+  }
+
+  SecKeychainItemRef item = nullptr;
+  OSQUERY_USE_DEPRECATED(while (SecKeychainSearchCopyNext(search, &item) == 0) {
+    SecKeychainAttributeList *attr_list = nullptr;
+
+    KeychainItemMetadata item_metadata;
+    item_metadata.keychain_path = path;
+
+    SecItemClass item_class = 0;
+    SecKeychainItemCopyAttributesAndData(
+        item, NULL, &item_class, NULL, NULL, NULL);
+
+    SecAccessRef access = nullptr;
+    os_status = SecKeychainItemCopyAccess(item, &access);
+    if (os_status == errSecNoAccessForItem) {
+      continue;
+    }
+
+    UInt32 item_id;
+    switch (item_class) {
+    case kSecInternetPasswordItemClass:
+      item_id = CSSM_DL_DB_RECORD_INTERNET_PASSWORD;
+      break;
+    case kSecGenericPasswordItemClass:
+      item_id = CSSM_DL_DB_RECORD_GENERIC_PASSWORD;
+      break;
+    case 'ashp':
+      item_id = CSSM_DL_DB_RECORD_APPLESHARE_PASSWORD;
+      break;
+    default:
+      item_id = item_class;
+      break;
+    }
+
+    SecKeychainAttributeInfo *info = nullptr;
+    SecKeychainAttributeInfoForItemID(keychain, item_id, &info);
+
+    os_status = SecKeychainItemCopyAttributesAndData(
+        item, info, &item_class, &attr_list, nullptr, nullptr);
+    if (os_status != 0) {
+      return Status(os_status,
+                    "Could not copy attributes and data from the keychain");
+    }
+
+    if (info == nullptr || attr_list == nullptr) {
+      return Status(1, "Could not get info or attr_list");
+    }
+
+    if (info->count != attr_list->count) {
+      return Status(1, "Info and attributes don't match");
+    }
+
+    for (int i = 0; i < info->count; ++i) {
+      SecKeychainAttribute *attribute = &attr_list->attr[i];
+      if (attribute->length == 0) {
+        continue;
+      }
+
+      UInt32 tag = info->tag[i];
+      if (tag == 7) {
+        item_metadata.label =
+            attributeBufferToString(attribute->data, attribute->length);
+      }
+    }
+
+    std::vector<KeychainItemACL> acl;
+    auto s = parseKeychainItemACL(access, acl);
+    if (!s.ok()) {
+      return s;
+    }
+
+    for (const auto &acl_data : acl) {
+      for (const auto &app_path : acl_data.applications) {
+        Row r;
+        r["keychain_path"] = item_metadata.keychain_path;
+        r["label"] = item_metadata.label;
+        r["authorizations"] = join(acl_data.authorizations, " ");
+        r["description"] = acl_data.description;
+        r["app_path"] = app_path;
+        results.push_back(r);
+      }
+    }
+  });
+
+  if (keychain != nullptr) {
+    CFRelease(keychain);
+  }
+  if (search != nullptr) {
+    CFRelease(search);
+  }
+  if (item != nullptr) {
```

Comment:
I think this check for item needs to go in to the while loop? You're "copy"ing during each iteration right?

There are several return/continue/break conditions within that loop too. I think the above releases may be avoided. It might be an easier read to break that loop logic into an inline/static method. You can control the allocs/releases outside of the inline method and simply check for status returns.